### DISCUSSION TITLE
Extensible mathematical generic word

### DISCUSSION THREAD

**Discussion by kusumotonorio** [factor/factor#2423] [Ideas]
This is a brainstorming session. :-)

One of the reasons for the desire to implement multiple methods in Factor is to achieve operator overloading. For example, [the first post](https://github.com/factor/factor/issues/2364#issue-737282625) [here](https://github.com/factor/factor/issues/2364) (#2364) , which I am now writing frequently, and I can see that being a request [here](https://github.com/factor/factor/issues/2261) (#2261) as well.

In a multi-method implementation such as the prototype I just presented, it is theoretically possible to write words such as `+` to be open to string concatenation, multi-dimensional array addition, etc., but if such operators were created as mathematical generic words, non-numeric parameters would not be available, and thus the implementation would not be able to use them, so it would not be usable for string concatenation in reality. In other words, the situation would be the same as it is now.

But...

If a mathematical generic word with multi-method support works by switching the dispatch strategy to a normal multi-method one, instead of raising an exception when given a non-numeric parameter load, then we can keep the advantages of the mathematical generic word and make it extensible Is that right? I don't think it's impossible.

**Comment by mrjbq7** [comment]
I think this would be a great improvement.

Not impossible!

In our current mathematical tree, we have "coercer" quotations that convert to make sure the arguments are both a specific type.  Perhaps in the "else" branch which right now throws, we allow overriding using multi-methods on both argument types?

**Reply by kusumotonorio** [reply]
I will try to think of ways to make this happen. I hope you will give it a try too.
Perhaps it could be extended without multi-method support?

**Comment by mrjbq7** [comment]
I wonder if providing means of adding coercer and types to the tree would be sufficient. Or provide the dispatch when the same type in the tree and error otherwise. That could work with the current method.

**Comment by kusumotonorio** [comment]
Apparently, it is working as expected. At least in the following example.

```factor
IN: scratchpad 
USE: mm9
 
MGENERIC: my-plus ( a b -- c ) mathematical

USING: math.private ;
MM: my-plus ( a:fixnum b: fixnum -- c ) fixnum+ ;
MM: my-plus ( a: float b: float -- c ) float+ ;

MM: my-plus ( a: string b: string -- c ) append ;

1 2.0 my-plus
"1 + 2 = " "3.0" my-plus
Loading resource:work/mm9/mm9.factor
Loading resource:basis/shuffle/shuffle.factor

--- Data stack:
3.0
"1 + 2 = 3.0"
IN: scratchpad 
```

**Comment by kusumotonorio** [comment]
I've added a few more.

```factor
IN: scratchpad 
USE: mm9
 
MGENERIC: my-plus ( a b -- c ) mathematical

USING: math.private ;
MM: my-plus ( a:fixnum b: fixnum -- c ) fixnum+ ;
MM: my-plus ( a: float b: float -- c ) float+ ;

MM: my-plus ( a: string b: string -- c ) append ;

USE: unicode
MM: my-plus ( a: character b: string -- c ) swap 1array >string prepend ;

1 2.0 my-plus
"1 + 2 = " "3.0" my-plus
CHAR: A "BC" my-plus

Loading resource:work/mm9/mm9.factor
Loading resource:basis/shuffle/shuffle.factor

--- Data stack:
3.0
"1 + 2 = 3.0"
"ABC"
IN: scratchpad 
```

**Reply by mrjbq7** [reply]
In this example, does it call the correct method with:

```factor
1 2 >bignum my-plus
1.0 2 my-plus
```

**Reply by kusumotonorio** [reply]
I can run the lower part with the current definition, but I need to add a definition for the upper part. I actually tried it.

```factor
IN: scratchpad MM: my-plus ( a: bignum b: bignum -- c ) bignum+ ;
1 2 >bignum my-plus dup class-of
1.0 2 my-plus

--- Data stack:
3
bignum
3.0
IN: scratchpad 
```

I was able to achieve the extensible mathematical generic word in a surprisingly short time, but I will pursue a better way of writing it.

**Reply by kusumotonorio** [reply]
As a side note, the current version automatically upgrades stack parameters if the two stack parameters are both in the numeric class, and does the usual multi-dispatch if either or both are not in the numeric class.
